File: php-app/api/imoveis.php
<?php
// ==========================================
// API REST - IMÓVEIS
// ==========================================
require_once '../config/database.php';
setJsonHeaders();

$method = $_SERVER['REQUEST_METHOD'];
$action = $_GET['action'] ?? '';

try {
    $db = getDB();
    
    switch ($method) {
        case 'GET':
            if ($action === 'all') {
                // Listar todos os imóveis ativos
                $stmt = $db->prepare("
                    SELECT i.*, c.nome as cliente_nome, c.cpf as cliente_cpf
                    FROM imoveis i
                    LEFT JOIN clientes c ON i.cliente_uuid = c.uuid
                    WHERE i.ativo = 1
                    ORDER BY i.data_cadastro DESC
                ");
                $stmt->execute();
                jsonResponse($stmt->fetchAll());
                
            } elseif ($action === 'get' && isset($_GET['uuid'])) {
                // Buscar imóvel por UUID
                $stmt = $db->prepare("
                    SELECT i.*, c.nome as cliente_nome, c.cpf as cliente_cpf
                    FROM imoveis i
                    LEFT JOIN clientes c ON i.cliente_uuid = c.uuid
                    WHERE i.uuid = ?
                ");
                $stmt->execute([$_GET['uuid']]);
                $imovel = $stmt->fetch();
                
                if ($imovel) {
                    jsonResponse($imovel);
                } else {
                    jsonResponse(['error' => 'Imóvel não encontrado'], 404);
                }
                
            } elseif ($action === 'by-cliente' && isset($_GET['cliente_uuid'])) {
                // Buscar imóveis por cliente
                $stmt = $db->prepare("
                    SELECT * FROM imoveis
                    WHERE cliente_uuid = ? AND ativo = 1
                    ORDER BY data_cadastro DESC
                ");
                $stmt->execute([$_GET['cliente_uuid']]);
                jsonResponse($stmt->fetchAll());
                
            } elseif ($action === 'by-tipo' && isset($_GET['tipo'])) {
                // Buscar imóveis por tipo
                $stmt = $db->prepare("
                    SELECT i.*, c.nome as cliente_nome
                    FROM imoveis i
                    LEFT JOIN clientes c ON i.cliente_uuid = c.uuid
                    WHERE i.tipo = ? AND i.ativo = 1
                    ORDER BY i.data_cadastro DESC
                ");
                $stmt->execute([$_GET['tipo']]);
                jsonResponse($stmt->fetchAll());
                
            } elseif ($action === 'search' && isset($_GET['termo'])) {
                // Buscar imóveis por termo
                $termo = '%' . $_GET['termo'] . '%';
                $stmt = $db->prepare("
                    SELECT i.*, c.nome as cliente_nome
                    FROM imoveis i
                    LEFT JOIN clientes c ON i.cliente_uuid = c.uuid
                    WHERE i.ativo = 1 AND (i.endereco LIKE ? OR c.nome LIKE ?)
                    ORDER BY i.data_cadastro DESC
                ");
                $stmt->execute([$termo, $termo]);
                jsonResponse($stmt->fetchAll());
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        case 'POST':
            if ($action === 'create') {
                // Criar novo imóvel
                $data = json_decode(file_get_contents('php://input'), true);
                
                if (!isset($data['tipo']) || !isset($data['endereco']) || 
                    !isset($data['valor_aluguel']) || !isset($data['contrato_meses']) || 
                    !isset($data['data_inicio']) || !isset($data['cliente_uuid'])) {
                    jsonResponse(['error' => 'Campos obrigatórios: tipo, endereco, valor_aluguel, contrato_meses, data_inicio, cliente_uuid'], 400);
                }
                
                $uuid = gerarUUID();
                $stmt = $db->prepare("
                    INSERT INTO imoveis (uuid, tipo, endereco, valor_aluguel, contrato_meses, data_inicio, cliente_uuid, periodo_arrendado, residencial, ativo)
                    VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, 1)
                ");
                $stmt->execute([
                    $uuid,
                    $data['tipo'],
                    $data['endereco'],
                    $data['valor_aluguel'],
                    $data['contrato_meses'],
                    $data['data_inicio'],
                    $data['cliente_uuid'],
                    $data['periodo_arrendado'] ?? null,
                    $data['residencial'] ?? null
                ]);
                
                jsonResponse(['success' => true, 'uuid' => $uuid], 201);
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        case 'PUT':
            if ($action === 'update' && isset($_GET['uuid'])) {
                // Atualizar imóvel
                $data = json_decode(file_get_contents('php://input'), true);
                $uuid = $_GET['uuid'];
                
                $campos = [];
                $valores = [];
                
                $camposPermitidos = ['tipo', 'endereco', 'valor_aluguel', 'contrato_meses', 'data_inicio', 'cliente_uuid', 'periodo_arrendado', 'residencial'];
                
                foreach ($camposPermitidos as $campo) {
                    if (isset($data[$campo])) {
                        $campos[] = "$campo = ?";
                        $valores[] = $data[$campo];
                    }
                }
                
                if (empty($campos)) {
                    jsonResponse(['error' => 'Nenhum campo para atualizar'], 400);
                }
                
                $valores[] = $uuid;
                $sql = "UPDATE imoveis SET " . implode(', ', $campos) . " WHERE uuid = ?";
                $stmt = $db->prepare($sql);
                $stmt->execute($valores);
                
                jsonResponse(['success' => true]);
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        case 'DELETE':
            if ($action === 'delete' && isset($_GET['uuid'])) {
                // Soft delete (desativar)
                $uuid = $_GET['uuid'];
                $stmt = $db->prepare("UPDATE imoveis SET ativo = 0 WHERE uuid = ?");
                $stmt->execute([$uuid]);
                
                jsonResponse(['success' => true]);
                
            } else {
                jsonResponse(['error' => 'Ação inválida'], 400);
            }
            break;
            
        default:
            jsonResponse(['error' => 'Método não permitido'], 405);
    }
    
} catch (Exception $e) {
    jsonResponse(['error' => $e->getMessage()], 500);
}
?>

File: python-app/routes/imoveis.py
from flask import Blueprint, request, jsonify
from models import Imovel, Cliente
from app import db
from datetime import datetime

imoveis_bp = Blueprint('imoveis', __name__)

@imoveis_bp.route('/', methods=['GET'])
def get_imoveis():
    """Listar todos os imóveis ativos"""
    imoveis = Imovel.query.filter_by(ativo=True).order_by(Imovel.data_cadastro.desc()).all()
    return jsonify([i.to_dict() for i in imoveis])

@imoveis_bp.route('/<uuid>', methods=['GET'])
def get_imovel(uuid):
    """Buscar imóvel por UUID"""
    imovel = Imovel.query.filter_by(uuid=uuid).first()
    if not imovel:
        return jsonify({'error': 'Imóvel não encontrado'}), 404
    return jsonify(imovel.to_dict())

@imoveis_bp.route('/cliente/<cliente_uuid>', methods=['GET'])
def get_imoveis_by_cliente(cliente_uuid):
    """Buscar imóveis por cliente"""
    imoveis = Imovel.query.filter_by(cliente_uuid=cliente_uuid, ativo=True).order_by(Imovel.data_cadastro.desc()).all()
    return jsonify([i.to_dict() for i in imoveis])

@imoveis_bp.route('/tipo/<tipo>', methods=['GET'])
def get_imoveis_by_tipo(tipo):
    """Buscar imóveis por tipo"""
    if tipo not in ['casa', 'fazenda', 'suite']:
        return jsonify({'error': 'Tipo inválido'}), 400
    
    imoveis = Imovel.query.filter_by(tipo=tipo, ativo=True).order_by(Imovel.data_cadastro.desc()).all()
    return jsonify([i.to_dict() for i in imoveis])

@imoveis_bp.route('/search', methods=['GET'])
def search_imoveis():
    """Buscar imóveis por termo"""
    termo = request.args.get('termo', '')
    imoveis = Imovel.query.join(Cliente).filter(
        Imovel.ativo == True,
        db.or_(
            Imovel.endereco.ilike(f'%{termo}%'),
            Cliente.nome.ilike(f'%{termo}%')
        )
    ).order_by(Imovel.data_cadastro.desc()).all()
    return jsonify([i.to_dict() for i in imoveis])

@imoveis_bp.route('/', methods=['POST'])
def create_imovel():
    """Criar novo imóvel"""
    data = request.get_json()
    
    required_fields = ['tipo', 'endereco', 'valor_aluguel', 'contrato_meses', 'data_inicio', 'cliente_uuid']
    if not data or not all(k in data for k in required_fields):
        return jsonify({'error': f'Campos obrigatórios: {", ".join(required_fields)}'}), 400
    
    # Verificar se cliente existe
    cliente = Cliente.query.filter_by(uuid=data['cliente_uuid'], ativo=True).first()
    if not cliente:
        return jsonify({'error': 'Cliente não encontrado'}), 404
    
    # Validar tipo
    if data['tipo'] not in ['casa', 'fazenda', 'suite']:
        return jsonify({'error': 'Tipo inválido'}), 400
    
    imovel = Imovel(
        tipo=data['tipo'],
        endereco=data['endereco'],
        valor_aluguel=float(data['valor_aluguel']),
        contrato_meses=int(data['contrato_meses']),
        data_inicio=datetime.strptime(data['data_inicio'], '%Y-%m-%d').date(),
        cliente_uuid=data['cliente_uuid'],
        periodo_arrendado=data.get('periodo_arrendado'),
        residencial=data.get('residencial')
    )
    
    db.session.add(imovel)
    db.session.commit()
    
    return jsonify({'success': True, 'uuid': imovel.uuid}), 201

@imoveis_bp.route('/<uuid>', methods=['PUT'])
def update_imovel(uuid):
    """Atualizar imóvel"""
    imovel = Imovel.query.filter_by(uuid=uuid).first()
    if not imovel:
        return jsonify({'error': 'Imóvel não encontrado'}), 404
    
    data = request.get_json()
    if not data:
        return jsonify({'error': 'Nenhum dado fornecido'}), 400
    
    # Atualizar campos
    campos_permitidos = ['tipo', 'endereco', 'valor_aluguel', 'contrato_meses', 'data_inicio', 
                         'cliente_uuid', 'periodo_arrendado', 'residencial']
    
    for campo in campos_permitidos:
        if campo in data:
            if campo == 'data_inicio':
                setattr(imovel, campo, datetime.strptime(data[campo], '%Y-%m-%d').date())
            elif campo in ['valor_aluguel', 'contrato_meses']:
                setattr(imovel, campo, float(data[campo]) if campo == 'valor_aluguel' else int(data[campo]))
            else:
                setattr(imovel, campo, data[campo])
    
    db.session.commit()
    
    return jsonify({'success': True})

@imoveis_bp.route('/<uuid>', methods=['DELETE'])
def delete_imovel(uuid):
    """Excluir imóvel (soft delete)"""
    imovel = Imovel.query.filter_by(uuid=uuid).first()
    if not imovel:
        return jsonify({'error': 'Imóvel não encontrado'}), 404
    
    imovel.ativo = False
    db.session.commit()
    
    return jsonify({'success': True})
